File: 8. 기타 그래프 이론/여행 계획.py
def find_parent(x):
    if parent[x] != x:
        parent[x] = find_parent(parent[x])
    return parent[x]

def union_parent(a, b):
    a = find_parent(a)
    b = find_parent(b)
    if a > b:
        parent[a] = b
    else:
        parent[b] = a

v, k = map(int, input().split())
parent = [i for i in range(v+1)]
for i in range(v):
  input_data = list(map(int, input().split()))
  for j in range(v):
    if input_data[j]:
      union_parent(i+1, j+1)

town = list(map(int, input().split()))
ans = 'YES'
for i in range(k-1):
  a = town[i]
  b = town[i+1]
  if find_parent(a) != find_parent(b):
    ans = 'NO'
    break

print(ans)

File: codeEg/8. 기타 그래프 이론/여행 계획.js
const input = `
0 1 0 1 1
1 0 1 1 0
0 1 0 0 0
1 1 0 0 0
1 0 0 0 0
`.trim().split('\n')

const [v, k] = [5, 4]
parent = new Array(v+1).fill().map((x,idx) => idx)

for (let i=0; i<v; i++) {
  tmp = input[i].split(' ').map(Number)
  for (let j=0; j<v; j++) {
    if (tmp[j]) {
      union_parent(i+1, j+1)
    }
  }
}

const town = [2, 3, 4, 3]
let ans = 'YES'
for (let i=0; i<k-1; i++) {
  const a = town[i]
  const b = town[i+1]
  if (find_parent(a) !== find_parent(b)) ans = 'NO'
}

console.log(ans)


function find_parent(x) {
  if (x != parent[x]) {
    parent[x] = find_parent(parent[x])
  }
  return parent[x]
}

function union_parent(a, b) {
  a = find_parent(a)
  b = find_parent(b)
  if (a > b) {
    parent[a] = b
  } else {
    parent[b] = a
  }
}